include/deal.II/differentiation/sd/symengine_scalar_operations.h

```diff
@@ -130,9 +135,676 @@ namespace Differentiation
+       *         overwhelming majority of cases, the default value of
+       *         <tt>false</tt> should be selected, with the result that an
+       *         exception will be thrown if the input @p symbolic is, in fact,
+       *         not a symbolic value or expression.
+       */
+      template <bool ignore_invalid_symbols = false>
+      void
+      add_to_substitution_map(types::substitution_map &       substitution_map,
+                              const SE::RCP<const SE::Basic> &symbol,
+                              const SE::RCP<const SE::Basic> &value);
+    } // namespace internal
+
+    /**
+     * A convenience function for an adding an entry to the substitution map
+     * @p substitution_map. The new entry will have the key given by @p symbol
+     * with its paired @p value. Such maps are required to perform substitution
+     * of symbolic expressions, with key entries being exchanges with their
+     * pair values.
+     *
+     * This function is guarenteed to create an ordering that is identical
+     * to the typical add_to_symbol_map() call that may used initially
+     * configuring a BatchOptimizer.
+     * This helps one conform to the requirement that the arguments sent into
+     * lambda and LLVM JIT compiled functions (created by optimizing symbolic
+     * expressions) (i) be the same, and (ii) have a constant ordering.
+     *
+     * @tparam ignore_invalid_symbols A template parameter that enforces whether
+     *         or not the @p symbol has to be a valid one or not. In the
+     *         overwhelming majority of cases, the default value of
+     *         <tt>false</tt> should be selected, with the result that an
+     *         exception will be thrown if the input @p symbolic is, in fact,
+     *         not a symbolic value or expression.
+     *         An exceptional case is, for example, when performing symbolic
+     *         assembly on a finite element level. When extracting the symbolic
+     *         equivalent of the shape function gradients using FEExtractors,
+     *         the returned tensor will have some <em>a priori</em> determined
+     *         zero-valued components. These trivial components are not valid
+     *         symbols (as they are not symbolic expressions), and we would
+     *         typically wish to guard against their (erroneous) inclusion. In
+     *         this scenario, for convenience, one could set
+     *         @p ignore_invalid_symbols to <tt>true</tt> and these zero-valued
+     *         entries would be skipped over and ignored.
+     *
+     * @note In this function, the @p ValueType is somewhat arbitrary as
+     * it is only used to create dummy values.
+     */
+    template <bool ignore_invalid_symbols = false>
+    void
+    add_to_substitution_map(types::substitution_map &substitution_map,
+                            const Expression &       symbol,
+                            const Expression &       value);
+
+    //    /**
```

Comment:
I was agonising as to whether to keep this or not. I've decided to remove these more "low level" functions, because the type that I was trying to support, namely `SE::RCP<const SE::Basic>`, can be converted into our `SD::Expression` anyway.